program-runtime/src/compute_budget.rs

```diff
@@ -122,25 +122,24 @@ impl ComputeBudget {
         }
     }
 
-    pub fn process_message(
+    pub fn process_instructions<'a>(
         &mut self,
-        message: &SanitizedMessage,
+        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
         requestable_heap_size: bool,
         default_units_per_instruction: bool,
     ) -> Result<u64, TransactionError> {
-        let mut num_instructions = message.instructions().len();
         let mut requested_additional_fee = 0;
         let mut requested_units = None;
-
-        let error = TransactionError::InstructionError(0, InstructionError::InvalidInstructionData);
-        for (i, (program_id, instruction)) in message.program_instructions_iter().enumerate() {
+        let mut num_instructions: usize = 0;
+        for (i, (program_id, instruction)) in instructions.enumerate() {
             if compute_budget::check_id(program_id) {
-                // don't include request instructions in default max calc
-                num_instructions = num_instructions.saturating_sub(1);
-
-                // Compute budget instruction must be in the 1st 3 instructions (avoid
-                // nonce marker), otherwise ignored
+                // Compute budget instruction must be in the 1st 3 instructions
+                // (need to account for nonce marker), otherwise ignored
                 if i < 3 {
+                    let error = TransactionError::InstructionError(
+                        i as u8,
+                        InstructionError::InvalidInstructionData,
+                    );
```

Comment:
Let's keep the original (wrong) error value for now that keeps `i` fixed at `0`. @jackcmay  has fixed this here: https://github.com/solana-labs/solana/pull/25027

Reply:
reverted to original `error`. Hopefully makes merge this function a bit easier later.